### ISSUE TITLE
Add LRU alike cache for SPIR-V to Metal shader source

### ISSUE THREAD

**Issue by Michael-Lfx** [gfx-rs/gfx#2494]
https://github.com/gfx-rs/gfx/blob/b9ebd07aa8b270c73a71d8e7fd59a06843d5c00c/src/backend/metal/src/device.rs#L1487-1500

Even though there is a todo for persistent AST to disk, but I think that adding a cache like LRU for mapping SPIR-V to Metal shader source is useful for Snapchat alike app which dynamically switches shaders on cellphone camera preview/recording.

**Comment by kvark** [comment]
Does the application try to create multiple shader modules from the same SPIR-V source? If that's the case, I don't think we should complicate the logic of the backend to handle this. Let's not try to be smarter than the user ;)

Otherwise, that LRU cache would have to be internally synchronized, e.g. wrapped in a `Mutex`. This way `create_shader_module` can continue to borrow `&self`.

**Comment by Michael-Lfx** [comment]
> Does the application try to create multiple shader modules from the same SPIR-V source?

Yes. Short video apps like Tik Tok can use one effect for a video clip. Usually users will record a lot of clips then merge them into the final 15s one. An effect contains multiple shader pairs. Most of them share the same vertex shader such as Beauty Face and Image Adjust, they owns different fragment shader to process specific image process algorithm. IMO, creating MSL from those vertex shader should be done once only because they are identical. There is another case. Users can switch different effects if they like. For example switching from A to B then A results in redundant SPIR-V to MSL and MSL compilation. I think there are chances for gfx-hal in the mobile video/image rendering situation.

**Comment by kvark** [comment]
@Michael-Lfx We aren't doing the SPIR-V to MSL translation at the `create_shader_module`, so putting LRU there will not get us anywhere at the moment. Problem is that the translation needs the resource binding overrides to be set, and those only depend on the pipeline layout. The solution to this I see is proper use of pipeline caches, which the implementation can re-use to avoid MSL generation. Are you using a pipeline cache?
